src/test/java/com/faforever/client/map/generator/MapGeneratorServiceTest.java

```diff
@@ -0,0 +1,148 @@
+    } catch (IOException e) {
+      e.printStackTrace();
+      throw new RuntimeException("Could not create test map generator folder!");
+    }
+
+    when(preferencesService.getFafDataDirectory()).thenReturn(fafDataDirectory.getRoot().toPath());
+
+    when(preferencesService.getPreferences()).thenReturn(preferences);
+    when(preferences.getForgedAlliance()).thenReturn(forgedAlliancePrefs);
+    when(forgedAlliancePrefs.getCustomMapsDirectory()).thenReturn(customMapsDirectory.getRoot().toPath());
+
+
+    instance = new MapGeneratorService(applicationContext, preferencesService, taskService);
+
+    assertThat(Arrays.asList(Objects.requireNonNull(customMapsDirectory.getRoot().list())), not(contains(testMapNameGenerator)));
+
+    when(downloadMapGeneratorTask.getFuture()).thenReturn(CompletableFuture.completedFuture(null));
+    when(generateMapTask.getFuture()).thenReturn(CompletableFuture.completedFuture(null));
+    doAnswer(invocation -> {
+      @SuppressWarnings("unchecked")
+      CompletableTask<Void> task = invocation.getArgument(0);
+      WaitForAsyncUtils.asyncFx(task);
+      task.getFuture().get();
+      return task;
+    }).when(taskService).submitTask(any());
+  }
+
+
+  @Test
+  public void testGenerateMapNoGeneratorPresent() {
+    when(applicationContext.getBean(DownloadMapGeneratorTask.class)).thenReturn(downloadMapGeneratorTask);
+    when(applicationContext.getBean(GenerateMapTask.class)).thenReturn(generateMapTask);
+
+    instance.generateMap(testMapNameNoGenerator).join();
+
+    verify(taskService).submitTask(downloadMapGeneratorTask);
+    verify(downloadMapGeneratorTask).setVersion(versionNoGeneratorPresent);
+
+    //See test below
+    verify(taskService).submitTask(generateMapTask);
+  }
+
+  @Test
+  public void testGenerateMapGeneratorPresent() {
+    when(applicationContext.getBean(GenerateMapTask.class)).thenReturn(generateMapTask);
+
+    instance.generateMap(testMapNameGenerator).join();
+
+    verify(taskService).submitTask(generateMapTask);
+    verify(generateMapTask).setVersion(versionGeneratorPresent);
+    verify(generateMapTask).setSeed(seed);
+    verify(generateMapTask).setGeneratorExecutableFile(fafDataDirectory.getRoot().toPath().resolve(MapGeneratorService.GENERATOR_EXECUTABLE_SUB_DIRECTORY).resolve(String.format(MapGeneratorService.getGENERATOR_EXECUTABLE_FILENAME(), versionGeneratorPresent)).toFile());
+    verify(generateMapTask).setMapFilename(testMapNameGenerator);
+
+    verifyNoMoreInteractions(taskService);
+  }
+
+  @Test
+  public void testWrongMapNameThrowsException() {
+    expectedException.expect(IllegalStateException.class);
+    expectedException.expectMessage(startsWith("No match found"));
+    CompletableFuture<String> future = instance.generateMap(testMapNameUnsupportedVersion);
+    future.join();
```

Comment:
You mean future.get right why do a join

Reply:
Ok it will have the same output

Reply:
No, I mean join, I'm not getting anything :)